java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,377 @@
+
+import javax.annotation.CheckForNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.model.JavaTree;
+import org.sonar.java.se.checkers.ConditionAlwaysTrueOrFalseChecker;
+import org.sonar.java.se.checkers.NullDereferenceChecker;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+public class ExplodedGraphWalker extends BaseTreeVisitor {
+
+  /**
+   * Arbitrary number to limit symbolic execution.
+   */
+  private static final int MAX_STEPS = 2000;
+  private static final Logger LOG = LoggerFactory.getLogger(ExplodedGraphWalker.class);
+  private static final Set<String> THIS_SUPER = ImmutableSet.of("this", "super");
+
+  private final ConditionAlwaysTrueOrFalseChecker alwaysTrueOrFalseChecker;
+  private ExplodedGraph explodedGraph;
+  private Deque<ExplodedGraph.Node> workList;
+  private ExplodedGraph.Node node;
+  public ExplodedGraph.ProgramPoint programPosition;
+  public ProgramState programState;
+
+  private CheckerDispatcher checkerDispatcher;
+
+  @VisibleForTesting
+  int steps;
+  ConstraintManager constraintManager;
+
+  public static class MaximumStepsReachedException extends RuntimeException {
+    public MaximumStepsReachedException(String s) {
+      super(s);
+    }
+  }
+
+  public ExplodedGraphWalker(JavaFileScannerContext context) throws MaximumStepsReachedException {
+    alwaysTrueOrFalseChecker = new ConditionAlwaysTrueOrFalseChecker();
+    this.checkerDispatcher = new CheckerDispatcher(this, context, Lists.newArrayList(alwaysTrueOrFalseChecker, new NullDereferenceChecker()));
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    super.visitMethod(tree);
+    BlockTree body = tree.block();
+    if (body != null) {
+      execute(tree);
+    }
+  }
+
+  private void execute(MethodTree tree) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "execute" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)